### ISSUE TITLE
[RESEARCH] Maximum PNG optimization

### ISSUE THREAD

**Issue by sergeevabc** [MrKrzYch00/zopfli#16]
Dear Krzysztof,
What is the proper way to achieve maximum PNG optimization/compression via this tool?
There are a lot of more or less documented switches, but not many usage examples.

**Comment by MrKrzYch00** [comment]
The more exhaustive search you run, better results you'll get. So running all possible switches, longest possible generic algorithm. And of course waiting for days if not months for single output.

In theory you should also run zopfli on every png test as gzip filter testing may, in some cases, provide worse results than it would be with zopfli for final zopfli part.

So for zopflipng:
- more genetic algorithm iterations or changing other of its settings may still give you some gains
- running zopfli instead of gzip may, in some scenarios, pick other filter that zopfli will end run will compress better, but the time waste makes this switch highly experimental

For zopfli (or zopfli compression in zopflipng, it is, however, recommended to extract IDAT and work on it separately):
- more iterations may still find better result when randomizing the data a bit basing on multiply with carry starting from predefined numbers from original zopfli which you can also feed differently through command line
- using --pass# may sometimes result in smaller file even with less iterations and the time used may be even less as it tries to find better split points for previously compressed stream (statistics change)
- using all compression modes will make zopfli mix them together, but it will take ~16x longer to compress

I mainly work with zopfli and first test each mix of 4 compression modes (16 combinations) to find best split points for lower number of iterations. Output the split point positions and feed zopfli with that split points again for 16 combinations while huffmix-ing the result (separate software) as well as using --all if there is 0-type block (as huffmix doesn't work with that). In all cases I use zopflidb cache to store some of statistics for skipping already processed data if the result would be the same (especially for --all last run as it will reuse some data).

In other words, not worth going for maximum compression. The best way is to test it out on various size files and pick some sane settings for each range of sizes. For example to never work longer than 1h / file. If you get smaller size than with other software that's something you should stay with and you can always come back to that one file and optimize further - run more genetic tests or/and various modes, more iterations in zopfli. This software just doesn't have limits on how long it can run, there may be still some very very small chance to get few bits less on 99999999 iteration. The RNG basing on feed should start repeating after 2,135,949,311 iteration I think as it uses 16 bit feeds... That would be possible maximum tries per block.

**Comment by sergeevabc** [comment]
Krzysztof, but what about a reasonable default “switches pack” to try Zopfli out of the box? Currently `zopflipng.exe in.png out.png` works too long and does unnecessary trials, even `-q` does not help. Big difference between that “unattainable unknown” and, for example, `pingo -s9 in.png` (delivers in seconds).

**Comment by MrKrzYch00** [comment]
This fork was not designed for speed but to add more control over the program with more code that may provide better result. Some optimizations were done, but this is only for zopfli compression. ZopfliPNG was mostly maintained to keep its interface to zopfli and retain most of its options in this fork.

On ZopfliPNG there were mostly other people working with their own forks and me including or merging their code (notably MegaByte). The way its working like this is as designed by others. However I can tell that most time consuming optimizations in ZopfliPNG come from incremental brute force and genetic algorithm (if run with a lot of iterations). Other factor is a lot of palette tries (if PNG is <=256 colors) which multiplies number of all tries with filters by amount of palette tries.
Incremental brute force works by adding line to line and recompressing the data while usual brute force works by compressing each line separately. The longer it works, the slower it gets.
However, both incremental brute force or/and genetic algorithm are usually winners.

As said before, due to optimizations being done mainly to zopfli (like multithreading support, which was never intended at first, mostly as a possible to do byproduct) extracting IDAT, recompressing it then injecting it back into PNG by retaining its main header data responsible for palette and filters is quite good solution. The IDAT is EXACTLY the same, just compressed better. Or 2 steps, first just PNG optimization with practically none zopfli compression then written previously zopfli tries separately (this is what I usually did).

If you build a house and expect every brick to be perfect, also expect a lot of time being spent on doing this. This is what those programs do, try to make it perfect or as close to perfection as possible without caring about how much time it costs.

**Comment by fhanau** [comment]
Aleksandr, I am the author of [ECT](https://github.com/fhanau/Efficient-Compression-Tool), which might be what you are looking for.
It is a file optimizer that is partially based on zopfli but runs a lot faster and usually offers a better compression ratio (depending on settings). While it offers fewer tuning options, I think it is good at offering reasonable default options and will usually outperform other software's compression ratio when using some of the more aggressive options.
